### DISCUSSION TITLE
Connecting to a language server via container networking

### DISCUSSION THREAD

**Discussion by krystalcode** [joaotavora/eglot#1470] [Q&A] [answered]
I run Emacs in a container, and the `phpactor` language server in another. I have the containers connected via a shared network.

I start the language server separately on port `8890` and it is listening for connections.

```
[LSP    ][INFO][916467.989963] Listening on 0.0.0.0:8890
```

When I ping `phpactor` from within the `emacs` container in a shell, the connection is established.

```
$ curl -v phpactor:8890
* Host phpactor:8890 was resolved.
* IPv6: (none)
* IPv4: 10.89.1.14, 10.89.1.14
*   Trying 10.89.1.14:8890...
* Connected to phpactor (10.89.1.14) port 8890
> GET / HTTP/1.1
> Host: phpactor:8890
> User-Agent: curl/8.9.1
> Accept: */*
> 
* Request completely sent off
```

And it reaches the server:

```
[LSP    ][INFO][916713.257669] Accepted connection from "10.89.1.15:48574"
[LSP    ][WARN][916713.262326] Could not parse header: GET / HTTP/1.1
[LSP    ][WARN][916713.262396] Could not parse header: Host: phpactor:8890
```

When I type `M-x eglot` I get prompted to type the host and port and I do so: `phpactor:8890`. I get the following error: "Searching for program: No such file or directory, phpactor:8890".

How do I get `eglot` to send the request to the given host/port instead of running a local program?

**Comment by joaotavora** [comment, marked as answer]
I can't reproduce this, but then again, I don't have half the picture.
Anyway, I think you might be using an old Eglot.  What is your Emacs
version?  Have you tried `M-x eglot-upgrade-eglot`?

On Fri, Nov 29, 2024 at 10:25 PM Dimitris Bozelos ***@***.***>
wrote:

> I run Emacs in a container, and the phpactor language server in another.
> I have the containers connected via a shared network.
>
> I start the language server separately on port 8890 and it is listening
> for connections.
>
> [LSP    ][INFO][916467.989963] Listening on 0.0.0.0:8890
>
> When I ping phpactor from within the emacs container in a shell, the
> connection is established.
>
> $ curl -v phpactor:8890
> * Host phpactor:8890 was resolved.
> * IPv6: (none)
> * IPv4: 10.89.1.14, 10.89.1.14
> *   Trying 10.89.1.14:8890...
> * Connected to phpactor (10.89.1.14) port 8890
> > GET / HTTP/1.1
> > Host: phpactor:8890
> > User-Agent: curl/8.9.1
> > Accept: */*
> >
> * Request completely sent off
>
> And it reaches the server:
>
> [LSP    ][INFO][916713.257669] Accepted connection from "10.89.1.15:48574"
> [LSP    ][WARN][916713.262326] Could not parse header: GET / HTTP/1.1
> [LSP    ][WARN][916713.262396] Could not parse header: Host: phpactor:8890
>
> When I type M-x eglot I get prompted to type the host and port and I do
> so: phpactor:8890. I get the following error: "Searching for program: No
> such file or directory, phpactor:8890".
>
> How do I get eglot to send the request to the given host/port instead of
> running a local program?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/joaotavora/eglot/discussions/1470>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC6PQ2ZCCXCZB7ARYFTLLD2DDSW3AVCNFSM6AAAAABSX2MX7WVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZXGU4DANJXGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
João Távora

**Reply by krystalcode** [reply]
Thank you for the swift response, almost figured it out. 

```
M-x emacs-version
GNU Emacs 29.4 (build 1, x86_64-redhat-linux-gnu, GTK+ Version 3.24.43, cairo version 1.18.0) of 2024-10-10
```

Looking for the Glot version, strangely `M-x package-list-packages` listed `eglot` as available, not installed. How come and I had all the `eglot` commands available, and it was working with a local `phpactor` executable? Anyways, maybe I messed something trying out different things.

I run `M-x eglot-upgrade-eglot` and version 1.17 is marked as installed. 

Now, having the following in my config

```
(add-to-list 'eglot-servers-program '((php-mode phps-mode) "phpactor" "8890"))
```

I run `M-x eglot` on a `php-mode` buffer, and it connects to the language server successfully. It does ask me to manually enter `host:port` every time though.

I want to do the following so that it connects automatically

```
(use-package eglot
  :config
  (add-to-list 'eglot-servers-program '((php-mode phps-mode) "phpactor" "8890"))
  :hook 
  (php-mode . eglot-ensure)
  (phps-mode . eglot-ensure))
```

Is `"phpactor" "8890"` not the correct syntax? [This page](https://www.gnu.org/software/emacs/manual/html_node/eglot/Setting-Up-LSP-Servers.html) states `(host port args…)`. Tried `"phpactor:8890"` as well, no luck. It fails (`[eglot] warning Searching for program: No such file or directory, phpactor:8890`), and I have to manually do `M-x eglot` and enter `phpactor:8890`, then it works.
